pkg/mcp/internal/ratelimiter.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+// RateLimit is partially representing standard lib's rate limiter
+type RateLimit interface {
+	Wait(ctx context.Context) (err error)
+}
+
+// ConnectionRateLimit is an interface for per connection rate limiter
+type ConnectionRateLimit interface {
+	Wait(ctx context.Context, id string) (err error)
+	Remove(id string)
+}
+
+// RateLimiter is wrapper around golang's rate.Limit
+// it associates each limiter to a unique connection ID
+// for more balanced rate limiting among multiple connections
+type RateLimiter struct {
+	connectionFreq      time.Duration
+	connectionBurstSize int
+	perConnRateLimiter  map[string]RateLimit
+}
+
+// NewRateLimiter returns a new RateLimiter
+func NewRateLimiter(freq time.Duration, burstSize int) *RateLimiter {
+	return &RateLimiter{
+		connectionFreq:      freq,
+		connectionBurstSize: burstSize,
+		perConnRateLimiter:  make(map[string]RateLimit),
+	}
+}
+
+// Wait associtates a limiter to a Connection ID
+// and calls wait with a given context
+func (c *RateLimiter) Wait(ctx context.Context, id string) error {
+	if limiter, ok := c.perConnRateLimiter[id]; ok {
```

Comment:
`Wait()` will be invoked by multiple goroutines, therefore `perConnRateLimiter` should be lock protected. This may create lock contention if _all_ incoming requests hit this same spot. Alternatively, you could refactor `ConnectionRateLimit` into a function/interface that returns the `RateLimit` interface. The caller could then decide to create a decide how it wants to manage the pool of limiters.

Reply:
Thanks for clarifying, I wasn't 100% sure how `EstablishResourceStream` (who is currently calling wait) was going to be used since there is no usage of it right now.

Reply:
`EstablishResourceStream` is invoked by the grpc server layer when a new client connects, e.g. https://github.com/istio/istio/blob/92f3d9716f1bf6f35befc07fd5deb1974b381f16/galley/pkg/server/server.go#L189

Reply:
@ayj 
>Alternatively, you could refactor `ConnectionRateLimit` into a function/interface that returns the `RateLimit` interface. The caller could then decide to create a decide how it wants to manage the pool of limiters.

I can't see how that would prevent the lock contention? either way a pool of limiters need to be managed and protected by mutex. Either by caller or the callee, any thoughts?

Reply:
I don't think we necessarily need an explicit pool. You could create a new rate limiter and later the golang runtime clean it up when its no longer referenced.

Reply:
I see what you mean now, done.